src/main.cpp

```diff
@@ -4623,6 +4624,17 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             return false;
         }
 
+        // When Overwinter is active, reject incoming connections from non-Overwinter nodes
+        if (NetworkUpgradeActive(GetHeight(), chainparams.GetConsensus(), Consensus::UPGRADE_OVERWINTER)
```

Comment:
`GetHeight() + 1` to address the off-by-one error.

Reply:
I don't think there's an off-by-one error.  At block height 9, Sprout rules still apply, thus a Sprout node should be allowed to connect; perhaps the node is syncing and makes a request for block data.

Reply:
It's a tricky one: Sprout rules apply to block height 9 itself, but Overwinter rules apply to blocks after height 9. So it depends on whether you consider Overwinter to be "active" once the last Sprout block has been received, or only once the first Overwinter block has been received. I think the latter is clearly incorrect (because the mempool needs to enforce Overwinter rules at height 9 in preparation for block 10), but I'm not sure the former is obviously or intuitively correct.

I think it makes sense to split the concept of "active-ness" into two categories:
- Consensus rules (obviously Overwinter rules apply only when block 10 is received)
- Non-consensus behaviours (at height 9 mempool only accepts Overwinter transactions, wallet will create Overwinter transactions, etc.)

I think the network behaviour falls into the latter category.

This affects #2808 because it needs to provide the status of each upgrade in a way that users of `getblockchaininfo` can understand and act on. If I were to think about a third party making transactions, or deciding at what point to start doing Overwinter things, I think I'd want to have an indication of the latter? Or more specifically, I'd want to have "Overwinter is active" occur right at the point when the block at height 9 has been fully-validated and accepted. This doesn't mesh with the way statuses are currently defined by `NetworkUpgradeStatus`, which assumes that "Overwinter is active" occurs right at the point when the block at height 10 has been fully-validated and accepted (which is when `chainActive.Height()` will start returning 10).

Reply:
In general we should consider Overwinter "active" once the last possible Sprout height has been reached.

Reply:
ZIP 200 states:

> ACTIVATION_HEIGHT
>
> The non-zero block height at which the network upgrade rules will come
> into effect, and be enforced as part of the blockchain consensus.
>
> For removal of ambiguity, the block at height ACTIVATION_HEIGHT - 1 is
> subject to the pre-upgrade consensus rules, and would be the last common
> block in the event of a persistent pre-upgrade branch.

If we consider a network upgrade to be "active" once the last possible height for the previous branch is reached, I think this would become:

> ACTIVATION_HEIGHT
>
> The non-zero block height at which the network upgrade rules will come
> into effect, and be enforced as part of the blockchain consensus.
>
> For removal of ambiguity, the block at height `ACTIVATION_HEIGHT` is itself
> subject to the pre-upgrade consensus rules, and would be the last common
> block in the event of a persistent pre-upgrade branch. The block at height
> `ACTIVATION_HEIGHT + 1` will be the first block to have the network upgrade
> rules enforced.

Is this a more sensible and/or understandable definition? Which one is prone to less confusion?

Reply:
Yeah, seems NetworkUpgradeState(upgrade) should return a value according to whether the *next* block is in the range [L,R)  where L=activation height of upgrade, R =activation height of next upgrade.

Reply:
Ohh syncing problems my "yeah" was to your previous comment and Sean's. Regarding your last comment, I don't think the definition of ACTIVATION_HEIGHT should change,  but the definition of when the upgrade is active should be according to what I wrote.

Reply:
> In general we should consider Overwinter "active" once the last possible Sprout height has been reached.

Can it be "active" if the activation height block never arrives?  Someone looking at the chain would see 100% Sprout blocks.

I think the ZIP is fine as is; perhaps add a paragraph distinguishing between the consensus rule _activation_ event when the new block arrives, and when a node should _alter its behaviour_ to prepare for that event.

We know that at `ACTIVATION_HEIGHT - 1` a node (can|should) change its behaviour to ensure a user action is executed as intended.  Creating a transaction will be version 3 so it can get into the next block.  Rejecting version 1 and 2 transactions from the mempool helps miners using getblocktemplate.

So should Overwinter nodes drop Sprout nodes at the `ACTIVATION_HEIGHT` or in the block before it?  Currently at the activation height we (1) reject incoming sprout connections (2) disconnect existing sprout connections when a message is sent, and before the activation height, we have some code which tries to maintain connections with Overwinter nodes rather than Sprout nodes in the run-up to activation, by prioritizing Sprout nodes for eviction.